platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetEventCreator.java

```diff
@@ -0,0 +1,371 @@
+            }
+        }
+
+        final EventDescriptor descriptor = buildDescriptor(event);
+        final List<EventDescriptor> parentDescriptors = getParentDescriptors(event);
+
+        tipsetBuilder.addEvent(descriptor, parentDescriptors);
+
+        if (!selfEvent) {
+            childlessOtherEventTracker.addEvent(descriptor, parentDescriptors);
+        }
+    }
+
+    /**
+     * Update the minimum generation non-ancient.
+     *
+     * @param minimumGenerationNonAncient the new minimum generation non-ancient
+     */
+    public void setMinimumGenerationNonAncient(final long minimumGenerationNonAncient) {
+        tipsetBuilder.setMinimumGenerationNonAncient(minimumGenerationNonAncient);
+        childlessOtherEventTracker.pruneOldEvents(minimumGenerationNonAncient);
+    }
+
+    /**
+     * Create a new event if it is legal to do so.
+     *
+     * @return the new event, or null if it is not legal to create a new event
+     */
+    @Nullable
+    public GossipEvent maybeCreateEvent() {
+        final long bullyScore = tipsetScoreCalculator.getBullyScore();
+        tipsetMetrics.getBullyScoreMetric().update(bullyScore);
+
+        // Never bother with anti-bullying techniques if we have a bully score of 1. We are pretty much guaranteed
+        // to bully ~1/3 of other nodes by a score of 1.
+        final double beNiceToNerdChance = (bullyScore - 1) / antiBullyingFactor;
+
+        if (beNiceToNerdChance > 0 && random.nextDouble() < beNiceToNerdChance) {
+            return createEventToReduceBullyScore();
+        } else {
+            return createEventByOptimizingTipsetScore();
+        }
+    }
+
+    /**
+     * Create an event using the other parent with the best tipset score.
+     *
+     * @return the new event, or null if it is not legal to create a new event
+     */
+    @Nullable
+    private GossipEvent createEventByOptimizingTipsetScore() {
+        final List<EventDescriptor> possibleOtherParents = childlessOtherEventTracker.getChildlessEvents();
+        Collections.shuffle(possibleOtherParents, random);
+
+        EventDescriptor bestOtherParent = null;
+        long bestScore = 0;
+        for (final EventDescriptor otherParent : possibleOtherParents) {
+            final long parentScore = tipsetScoreCalculator.getTheoreticalAdvancementScore(List.of(otherParent));
+            if (parentScore > bestScore) {
+                bestOtherParent = otherParent;
+                bestScore = parentScore;
+            }
+        }
+
+        if (lastSelfEvent != null && bestOtherParent == null) {
+            // There exist no parents that can advance consensus, and this is not our first event.
+            return null;
+        }
+
+        return buildEvent(bestOtherParent);
+    }
+
+    /**
+     * Create an event that reduces the bully score.
+     *
+     * @return the new event, or null if it is not legal to create a new event
+     */
+    @Nullable
+    private GossipEvent createEventToReduceBullyScore() {
```

Comment:
intellij code coverage shows this method as never being called. Are there any test changes you can make so that events are made to intentionally reduce bully score?

Reply:
I will work on adding a unit test that executes this.